packages/openai-compatible/src/openai-compatible-provider.ts

```diff
@@ -0,0 +1,183 @@
+import {
+  EmbeddingModelV1,
+  LanguageModelV1,
+  ProviderV1,
+} from '@ai-sdk/provider';
+import {
+  FetchFunction,
+  loadApiKey,
+  withoutTrailingSlash,
+} from '@ai-sdk/provider-utils';
+import { OpenAICompatibleChatLanguageModel } from './openai-compatible-chat-language-model';
+import { OpenAICompatibleChatSettings } from './openai-compatible-chat-settings';
+import { OpenAICompatibleCompletionLanguageModel } from './openai-compatible-completion-language-model';
+import { OpenAICompatibleCompletionSettings } from './openai-compatible-completion-settings';
+import { OpenAICompatibleEmbeddingSettings } from './openai-compatible-embedding-settings';
+import { OpenAICompatibleEmbeddingModel } from './openai-compatible-embedding-model';
+
+export interface OpenAICompatibleProvider<
+  CHAT_MODEL_IDS extends string = string,
+  COMPLETION_MODEL_IDS extends string = string,
+  EMBEDDING_MODEL_IDS extends string = string,
+> extends ProviderV1 {
+  (
+    modelId: CHAT_MODEL_IDS,
+    settings?: OpenAICompatibleChatSettings,
+  ): LanguageModelV1;
+
+  languageModel(
+    modelId: CHAT_MODEL_IDS,
+    settings?: OpenAICompatibleChatSettings,
+  ): LanguageModelV1;
+
+  chatModel(
+    modelId: CHAT_MODEL_IDS,
+    settings?: OpenAICompatibleChatSettings,
+  ): LanguageModelV1;
+
+  completionModel(
+    modelId: COMPLETION_MODEL_IDS,
+    settings?: OpenAICompatibleCompletionSettings,
+  ): LanguageModelV1;
+
+  textEmbeddingModel(
+    modelId: EMBEDDING_MODEL_IDS,
+    settings?: OpenAICompatibleEmbeddingSettings,
+  ): EmbeddingModelV1<string>;
+}
+
+export interface OpenAICompatibleProviderSettings {
+  /**
+Base URL for the API calls.
+     */
+  baseURL?: string;
+
+  /**
+API key for authenticating requests.
+     */
+  apiKey?: string;
+
+  /**
+Custom headers to include in the requests.
+     */
+  headers?: Record<string, string>;
+
+  /**
+Custom fetch implementation. You can use it as a middleware to intercept requests,
+or to provide a custom fetch implementation for e.g. testing.
+    */
+  fetch?: FetchFunction;
+
+  /**
+The name of the environment variable from which to load the API key (if a key isn't explicitly provided).
+   */
+  apiKeyEnvVarName?: string;
+
+  /**
+Description of the API key environment variable (for use in error messages).
+   */
+  apiKeyEnvVarDescription?: string;
+
```

Comment:
lets remove api key related code completely. instead users can provide the key via headers. reason being that some providers have no api keys and there are different ways to send it via headers